evm/src/proof.rs

```diff
@@ -123,9 +196,40 @@ pub struct BlockMetadata {
     pub block_bloom: [U256; 8],
 }
 
+impl BlockMetadata {
+    pub fn from_public_inputs<F: RichField>(pis: &[F]) -> Self {
+        assert!(pis.len() == BlockMetadataTarget::SIZE);
+
+        let block_beneficiary = get_h160(&pis[0..5]);
+        let block_timestamp = pis[5].to_canonical_u64().into();
+        let block_number = pis[6].to_canonical_u64().into();
+        let block_difficulty = pis[7].to_canonical_u64().into();
+        let block_random = get_h256(&pis[8..16]);
+        let block_gaslimit = pis[16].to_canonical_u64().into();
+        let block_chain_id = pis[17].to_canonical_u64().into();
+        let block_base_fee =
+            (pis[18].to_canonical_u64() + (pis[19].to_canonical_u64() << 32)).into();
+        let block_gas_used = pis[20].to_canonical_u64().into();
+        let block_bloom = core::array::from_fn(|i| h2u(get_h256(&pis[21 + 8 * i..29 + 8 * i])));
```

Comment:
These ranges are bit awkward and it looks a bit brittle. If one of us make a change here, would we ever notice? Can we at least move the extraction logic to `PublicValues`? Have `pis.block_beneficiary()` etc and hope the compiler optimizes it?

Reply:
Yeah I agree the range accesses are quite error-prone (we use the same in the target versions to construct the proofs, so this should normally be easily detected if there was a breaking change).
Ideally I'd like to review the way we construct those objects, maybe something similar to the different views (like `CpuColumnsView`?) to remove the need for numeric accesses.